subsys/bluetooth/services/ots_oacp.c

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/printk.h>
+#include <sys/byteorder.h>
+#include <zephyr.h>
+
+#include <bluetooth/gatt.h>
+#include <bluetooth/services/ots.h>
+#include "ots_internal.h"
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(bt_gatt_ots, CONFIG_BT_GATT_OTS_LOG_LEVEL);
+
+#define OACP_RES_MAX_SIZE 3
+
+static enum bt_gatt_ots_oacp_res_code oacp_read_proc_validate(
+	struct bt_conn *conn,
+	struct bt_gatt_ots *ots,
+	struct bt_gatt_ots_oacp_proc *proc)
+{
+	struct bt_gatt_ots_oacp_read_params *params = &proc->read_params;
+
+	LOG_DBG("Validating Read procedure with offset: 0x%08X and "
+		"length: 0x%08X", params->offset, params->len);
+
+	if (!ots->cur_obj) {
+		return BT_GATT_OTS_OACP_RES_INV_OBJ;
+	}
+
+	if (!ots->cur_obj->metadata.props.decoded.read_permitted) {
+		return BT_GATT_OTS_OACP_RES_NOT_PERMITTED;
+	}
+
+	if (!bt_gatt_ots_l2cap_is_open(&ots->l2cap, conn)) {
+		return BT_GATT_OTS_OACP_RES_CHAN_UNAVAIL;
+	}
+
+	if ((params->offset + (u64_t) params->len) >
+		ots->cur_obj->metadata.size.cur) {
+		return BT_GATT_OTS_OACP_RES_INV_PARAM;
+	}
+
+	memcpy(&ots->cur_proc, proc, sizeof(ots->cur_proc));
+
+	LOG_DBG("Read procedure is accepted");
+
+	return BT_GATT_OTS_OACP_RES_SUCCESS;
+}
+
+static enum bt_gatt_ots_oacp_res_code oacp_proc_validate(
+	struct bt_conn *conn,
+	struct bt_gatt_ots *ots,
+	struct bt_gatt_ots_oacp_proc *proc)
+{
+	switch (proc->type) {
+	case BT_GATT_OTS_OACP_PROC_READ:
+		return oacp_read_proc_validate(conn, ots, proc);
+	case BT_GATT_OTS_OACP_PROC_CREATE:
+	case BT_GATT_OTS_OACP_PROC_DELETE:
+	case BT_GATT_OTS_OACP_PROC_CHECKSUM_CALC:
+	case BT_GATT_OTS_OACP_PROC_EXECUTE:
+	case BT_GATT_OTS_OACP_PROC_WRITE:
+	case BT_GATT_OTS_OACP_PROC_ABORT:
+		return BT_GATT_OTS_OACP_RES_OPCODE_NOT_SUP;
+	default:
+		return BT_GATT_OTS_OACP_RES_INV_PARAM;
```

Comment:
Incorrect, this should be OPCODE Not supported (all values in the default here are RFU and the OTS specification stipulate in section 3.3.8.2.1 'This includes Op Code values which are reserved for future use.'